runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/SdkHarnessClient.java

```diff
@@ -310,6 +312,9 @@ public ActiveBundle newBundle(
       private final BundleSplitHandler splitHandler;
       private final BundleCheckpointHandler checkpointHandler;
       private final BundleFinalizationHandler finalizationHandler;
+      private final Phaser outstandingRequests;
+      private final AtomicBoolean isClosed;
+      private boolean bundleIsCompleted;
```

Comment:
Does this need to be `volatile`?

Reply:
Looks like the answer is no because `synchronized` blocks flush caches.

Reply:
no, its always used in a synchronized block.